swarm/storage/binarymerkle.go

```diff
@@ -0,0 +1,384 @@
+	//      label is hash of (children[0].label + children[1].label)
+	// if leaf: label is arbitrary data
+	// else if children[1] is nil, label=hash(children[0].label)
+}
+
+func (t BTree) Count() uint64 {
+	return t.count
+}
+
+// The hash/root of an empty BTree does not matter
+func (t BTree) Root() []byte {
+	return t.rootHash
+}
+
+// All trees should pass , unless they are invalid, which should only happen
+// if incorrectly built or modified.
+// Checks the rep invariants
+func (t BTree) Validate() error {
+	count, height, error := t.root.validate()
+	if error != nil {
+		return error
+	}
+	if count != t.count {
+		return fmt.Errorf("Incorrect count. Was %d, should be %d", t.count, count)
+	}
+	if height != GetHeight(count) {
+		return fmt.Errorf("Incorrect height. Was %d, should be %d", height, GetHeight(count))
+	}
+
+	rootLabel := make([]byte, 0)
+	if height > 0 {
+		rootLabel = t.root.label
+	}
+	h := rootHash(count, rootLabel)
+	if !bytes.Equal(t.rootHash, h) {
+		return fmt.Errorf("Incorrect rootHash")
+	}
+	return nil
+}
+
+// Checks the rep invariants
+func (t *node) validate() (count uint64, height int, err error) {
+	if t == nil {
+		return 0, 0, nil
+	}
+	if t.children[0] == nil {
+		if t.children[1] != nil {
+			return 0, 0, fmt.Errorf("Invalid Node: Node missing first child, but has second")
+		}
+		// Leaf node
+		return 1, 1, nil
+	}
+
+	// Not a leaf node
+	count, height, err = t.children[0].validate()
+	if err != nil {
+		return
+	}
+	if t.children[1] != nil {
+		count2, height2, err2 := t.children[1].validate()
+		count += count2
+		if err2 != nil {
+			return count, height, err2
+		}
+		if height2 != height {
+			return count, height, fmt.Errorf("Invalid Node: height mismatch between children")
+		}
+	}
+	h := makeHash(t.children[0], t.children[1])
+	if !bytes.Equal(h, t.label) {
+		return 0, 0, fmt.Errorf("Invalid Node: Node hash mismatch")
+	}
+
+	height++
+	return
+}
+
+func rootHash(count uint64, data []byte) []byte {
+	h := hashFunc()
+	h.Reset()
+	h.Write(data)
+	binary.Write(h, binary.LittleEndian, count)
+	return h.Sum(make([]byte, 0))
+}
+
+func makeHash(left, right *node) []byte {
+	h := hashFunc()
+	h.Reset()
+	if left != nil {
+		h.Write(left.label)
+		if right != nil {
+			h.Write(right.label)
+		}
+	}
+	return h.Sum(make([]byte, 0))
+}
+
+// Returns the height of the tree containing count leaf nodes.
+// This the number of nodes (including the final leaf) from the root to
+// any leaf.
+func GetHeight(count uint64) int {
+	if count == 0 {
+		return 0
+	}
+	height := 0
+	for count > (1 << uint(height)) {
+		height++
+	}
+	return height + 1
+}
+
+// Build Binary Merkle Tree over data segments of segmentsize len with a specific hash func
+// Return
+// BMT - The BMT Representation of the data
+// ROOT - BMT Root
+// Count - Numers of leafs at the BMT
+// error - if exist validation(-1) count(-2) ok(0)
+func BuildBMT(h Hasher, data []byte, segmentsize int) (bmt *BTree, roor *Root, count int, errorcode int) {
```

Comment:
The BMT will be calculated on max 4096 bytes.
Are you referring to run this BMT in parallel on multiple chunks ? Is it already been done by chunker? 
I will try to make the BuildBMT more efficient with parallel computations.
Will look on the pyramid for reference.